### PR TITLE
Track the stocks purchased by users

### PR DESCRIPTION
**Pull request by sumitd94**
This PR contains the logic to be able to store the stocks the user has.
> This is in continuation with this PR #222

### CODE REVIEW

File: routes/users.js [hunk 1]
```diff
@@ -209,4 +210,28 @@ router.get('/isUsernameAvailable/:username', authenticate, usersController.getUs
...
+ *             $ref: '#/components/schemas/userStocks'
+ *     500:
+ *       description: badImplementation
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/errors/badImplementation'
+ */
+
+router.get('/stocks/self', authenticate, getUserStocks)
```
Comment (ankushdharkar) [inline routes/users.js:235]: This should ideally be in a separate routes file

File: models/stocks.js [hunk 2]
```diff
@@ -38,7 +40,84 @@ const fetchStocks = async () => {
...
+      return {}
+    }
+
+    userStocksRef = await userStocksModel.where('userId', '==', userId).get()
+    const userStocks = []
+    userStocksRef.forEach((stock) => {
+      userStocks.push({
+        id: stock.id,
+        ...stock.data()
+      })
```
Comment (ankushdharkar) [inline models/stocks.js:65]: Lot of repetition, can this be simplified?

File: models/trading.js [hunk 3]
```diff
@@ -45,34 +47,53 @@ const trade = async (tradeData) => {
     switch (tradeType) {
       case 'SELL': {
         quantityToUpdate = quantity + stockData.quantity
-        userBalance = quantity * stockData.price
-        updatedCurrencyData.dinero = userBalance + currencies.dinero
-        stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+        userBalance = (quantity * stockData.price) + currencies.dinero
+        updatedCurrencyData.dinero = userBalance
+        userStocksQty = userStocks.quantity - quantity
```
Comment (ankushdharkar) [inline models/trading.js:52]: Where are we checking if it's allowed?

**Review by ankushdharkar — DISMISSED** [review]

File: models/stocks.js (hunk 2 above)
Comment (sumitd94) [reply models/stocks.js:65]: Yes, I am finding a way to making it simple.

File: routes/users.js (hunk 1 above)
Comment (sumitd94) [reply routes/users.js:235]: Ok Sure, will move it into `stocks` routes

File: models/trading.js (hunk 3 above)
Comment (sumitd94) [reply models/trading.js]: d0ce9ca

File: routes/users.js (hunk 1 above)
Comment (sumitd94) [reply routes/users.js]: d0ce9ca

File: models/stocks.js (hunk 2 above)
Comment (sumitd94) [reply models/stocks.js]: https://github.com/Real-Dev-Squad/website-backend/pull/265/commits/b2cd138a77efde53e934a063fd6f021608731fcf

File: models/trading.js [hunk 4]
```diff
@@ -26,16 +28,17 @@ const getUpdatedPrice = (stockPrice) => {
 const trade = async (tradeData) => {
```
Comment (ankurnarkhede) [inline models/trading.js:28]: @sumitd94 , can we have a tradeService.js and add all the business logic there, and keep only the CRUD logic in models files?

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: models/trading.js (hunk 4 above)
Comment (ankurnarkhede) [reply models/trading.js:28]: Similar to what we do for auth, github and contributions..

Comment (sumitd94) [reply models/trading.js:28]: The entire logic is business here, should I move all the logic into a service file and call this function from the controller?

Comment (sumitd94) [reply models/trading.js:28]: 2d78b75

File: models/stocks.js [hunk 5]
```diff
@@ -38,7 +40,85 @@ const fetchStocks = async () => {
...
+const createUserStock = async (userId, stockData) => {
+  try {
+    const userStocks = {
+      userId,
+      ...stockData
+    }
+    const { id } = await userStocksModel.add(userStocks)
+    return {
+      id
+    }
```
Comment (swarajpure) [inline models/stocks.js:88]: NIT: Can be condensed in one line

File: models/stocks.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 const firestore = require('../utils/firestore')
 const stocksModel = firestore.collection('stocks')
+const userStocksModel = firestore.collection('user-stocks')
```
Comment (swarajpure) [inline models/stocks.js:3]: We're using `stocks` right? The firestore collection

File: models/stocks.js [hunk 7]
```diff
@@ -38,7 +40,85 @@ const fetchStocks = async () => {
...
+      await createUserStock(userId, stockData)
+      return true
+    }
+
+    const userStocksRef = userStocksModel.doc(userStocks.id)
+    const res = await userStocksRef.update(stockData)
+    if (res) {
+      return true
+    }
+    return false
```
Comment (swarajpure) [inline models/stocks.js:112]: NIT: can be simplified

File: routes/stocks.js [hunk 8]
```diff
@@ -64,4 +64,28 @@ router.get('/', fetchStocks)
...
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/userStocks'
+ *     500:
+ *       description: badImplementation
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/errors/badImplementation'
+ */
```
Comment (swarajpure) [inline routes/stocks.js:87]: Missing 401 for unAuthorized

**Review by swarajpure — CHANGES_REQUESTED** [review]
Having some queries with the firestore data model, also some NIT comments.

File: models/stocks.js (hunk 6 above)
Comment (sumitd94) [reply models/stocks.js:3]: `stocks` collection only contains the details of the stocks which the users can see on screen and trade them.
`user-stocks` collection contains the details of the stocks the individual users have

The API contracts need to be updated for this, will be doing that soon.

File: routes/stocks.js (hunk 8 above)
Comment (sumitd94) [reply routes/stocks.js]: e33ae7a

File: models/stocks.js (hunk 7 above)
Comment (sumitd94) [reply models/stocks.js]: 5c69f89

File: models/stocks.js (hunk 5 above)
Comment (sumitd94) [reply models/stocks.js]: 5c69f89

File: services/tradingService.js [hunk 9]
```diff
@@ -44,35 +47,57 @@ const trade = async (tradeData) => {
...
       case 'SELL': {
+        if (!userStocks.id || userStocks.quantity < quantity) {
+          return { canUserTrade: false, errorMessage: INSUFFICIENT_QUANTITIES }
+        }
+
         quantityToUpdate = quantity + stockData.quantity
-        userBalance = quantity * stockData.price
-        updatedCurrencyData.dinero = userBalance + currencies.dinero
-        stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+        userBalance = (quantity * stockData.price) + currencies.dinero
```
Comment (swarajpure) [inline services/tradingService.js:55]: Please use constants for currency names -- dinero, neelam. We already have a constants file

File: services/tradingService.js [hunk 10]
```diff
@@ -44,35 +47,57 @@ const trade = async (tradeData) => {
...
         qtyUserCanPurchase = Math.floor(totalPrice / stockData.price)
         if (qtyUserCanPurchase <= 0 || totalPrice > currencies.dinero) {
           return { canUserTrade: false, errorMessage: INSUFFICIENT_FUNDS }
         }
-        orderValue = qtyUserCanPurchase * stockData.price
         quantityToUpdate = stockData.quantity - qtyUserCanPurchase
-        userBalance = currencies.dinero - orderValue
-        updatedCurrencyData.dinero = userBalance
-        stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+        userBalance = currencies.dinero - (qtyUserCanPurchase * stockData.price)
```
Comment (swarajpure) [inline services/tradingService.js:65]: re: constants

File: services/tradingService.js [hunk 11]
```diff
@@ -44,35 +47,57 @@ const trade = async (tradeData) => {
...
         break
       }
       default: {
         return { canUserTrade: false, errorMessage: 'Invalid trade type' }
       }
     }
 
+    const orderValue = qtyUserCanPurchase * stockData.price
+    const stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+    updatedCurrencyData.dinero = userBalance
```
Comment (swarajpure) [inline services/tradingService.js:83]: re: constants

File: models/stocks.js [hunk 12]
```diff
@@ -38,7 +40,80 @@ const fetchStocks = async () => {
...
 }
 
+/**
+ * Fetches the user stocks
+ * @return {Promise<userStocks|object>}
+ */
+const fetchUserStocks = async (userId, stockId = null) => {
+  try {
+    let userStocksRef = ''
+    const query = userStocksModel.where('userId', '==', userId)
```
Comment (swarajpure) [inline models/stocks.js:50]: What about the case when the given userId is not found? Example, If I search for stocks of Elon Musk

I understand this function is for self user, but someone in future will use this function to get someone else's stock and they may not consider this case when the user provided is wrong.

File: controllers/stocks.js [hunk 13]
```diff
@@ -37,8 +37,28 @@ const fetchStocks = async (req, res) => {
     return res.boom.badImplementation('An internal server error occurred')
   }
 }
+/**
+ * Fetches all the stocks of the user
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const getUserStocks = async (req, res, next) => {
```
Comment (swarajpure) [inline controllers/stocks.js:46]: Do we need `next`? I guess controller is the last middleware 🤔

**Review by swarajpure — CHANGES_REQUESTED** [review]
Requesting some changes, please check.

Thank you! 😄

File: models/stocks.js [hunk 14]
```diff
@@ -38,7 +40,80 @@ const fetchStocks = async () => {
...
+    logger.error('Error creating user stocks', err)
+    throw err
+  }
+}
+
+/**
+ * Update Users Stocks
+ * @return {Promise<userStocks|object>}
+ */
+const updateUserStocks = async (userId, stockData) => {
```
Comment (ankurnarkhede) [inline models/stocks.js:97]: We can simply create an `addOrUpdate` function here and remove the above function, as this function is also handling the adding stocks part.

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: models/stocks.js (hunk 14 above)
Comment (sumitd94) [reply models/stocks.js:97]: This function handles only updating of stocks. I created the above `createUserStocks` function just to make the code look cleaner and readable.

File: models/stocks.js (hunk 12 above)
Comment (sumitd94) [reply models/stocks.js:50]: I understand what you mean here and that makes sense, but I think when in the future we use this, we can add this feature .. what say?

File: models/stocks.js (hunk 14 above)
Comment (sumitd94) [reply models/stocks.js:79]: 4668438

File: models/stocks.js (hunk 12 above)
Comment (sumitd94) [reply models/stocks.js:50]: As per our discussion, closing this .. we do need this check here.

File: services/tradingService.js (hunk 9 above)
Comment (sumitd94) [reply services/tradingService.js]: a324cbc

**Review by ankurnarkhede — DISMISSED** [review]

**Review by swarajpure — DISMISSED** [review]
Looks good to me!

File: models/trading.js (hunk 4 above)
Comment (ankurnarkhede) [reply models/trading.js:29]: Yup 👍

File: controllers/stocks.js [hunk 15]
```diff
@@ -37,8 +37,28 @@ const fetchStocks = async (req, res) => {
...
+ * @param res {Object} - Express response object
+ */
+const getSelfStocks = async (req, res) => {
+  try {
+    const { id: userId } = req.userData
+    const userStocks = await stocks.fetchUserStocks(userId)
+    return res.json({
+      message: 'User stocks returned successfully!',
+      userStocks
+    })
```
Comment (swarajpure) [inline controllers/stocks.js:53]: Can we have the special message for the case when there `userStocks` is `[]` something like `No stocks found`? (Using ternary operator)

File: services/tradingService.js [hunk 16]
```diff
@@ -44,35 +48,57 @@ const trade = async (tradeData) => {
...
-        updatedCurrencyData.dinero = userBalance + currencies.dinero
-        stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+        userBalance = (quantity * stockData.price) + currencies[walletConstants.DINERO]
+        userStocksQty = userStocks.quantity - quantity
         break
       }
       case 'BUY': {
         qtyUserCanPurchase = Math.floor(totalPrice / stockData.price)
-        if (qtyUserCanPurchase <= 0 || totalPrice > currencies.dinero) {
+        if (qtyUserCanPurchase <= 0 || totalPrice > currencies[walletConstants.DINERO]) {
```
Comment (swarajpure) [inline services/tradingService.js:62]: I see the wallet constants being used a lot of times in this file. Can we destructure them and use?

**Review by swarajpure — CHANGES_REQUESTED** [review]
Suggesting some minor changes. Please have a look once.
Thank you!

File: controllers/stocks.js (hunk 15 above)
Comment (sumitd94) [reply controllers/stocks.js:53]: Is it really needed? I mean in all other places as well we are just doing this same thing.

```javascript
return res.json({
message: '<something> returned successfully!',
<whatever you are requesting (doesnt matter if its [] or [{...data}]>
})
```

File: services/tradingService.js (hunk 16 above)
Comment (sumitd94) [reply services/tradingService.js]: a383c4b

File: controllers/stocks.js (hunk 15 above)
Comment (swarajpure) [reply controllers/stocks.js:53]: If I have 0 zero stocks in my account, wouldn't the message `User stocks returned successfully!` confuse me? It says that it has returned stocks successfully but the stocks array is `[]`

Comment (sumitd94) [reply controllers/stocks.js]: 6310a35

**Review by swarajpure — APPROVED** [review]
Looks good to me!
Great job! ⚡

**Pull request merged**